### PR TITLE
week 6 assessment complete

### PR DESCRIPTION
**Pull request by don-yoon**
# Submitting Week 6 Assessment

### ✅ Completed Files
Select the following as confirmation that the files in this repository are complete.
- [x] Coding Challenges
- [x] Coding Concepts
- [x] Interview Questions

### 📝 Feedback
If you would like specific feedback on a particular problem please leave details here:

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -12,40 +12,124 @@
...
+      { name: "ford prefect", occupation: "a hitchhiker" },
+      { name: "zaphod beeblebrox", occupation: "president of the galaxy" },
+      { name: "arthur dent", occupation: "a radio employee" }
+    ]
+    // Expected output: ["Ford Prefect is a hitchhiker.", "Zaphod Beeblebrox is president of the galaxy.", "Arthur Dent is a radio employee."]
+    expect(aboutChar(people)).toEqual(["Ford Prefect is a hitchhiker.", "Zaphod Beeblebrox is president of the galaxy.", "Arthur Dent is a radio employee."])
+  })
+})
+
+//! ● aboutChar › takes in an array of objects and returns an array with a sentence about each person with their name capitalized.
```
Comment (SunkissedQueen) [inline code-challenges.test.js:29]: good fail

File: code-challenges.test.js [hunk 2]
```diff
@@ -12,40 +12,124 @@
...
+// }
+
+//* PASS  ./code-challenges.test.js
+//* aboutChar
+//*   ✓ takes in an array of objects and returns an array with a sentence about each person with their name capitalized. (2 ms)
+
+//TODO REFACTOR
+//? We don't need to store any of the map results into an array
+const aboutChar = array => array.map(value => `${value.name} is ${value.occupation}.`).map(value => value.replace(value[0], value[0].toUpperCase())).map(value => value.replace(value[value.indexOf(" ") + 1], value[value.indexOf(" ") + 1].toUpperCase()))
+
```
Comment (SunkissedQueen) [inline code-challenges.test.js:48]: nice approach
How would this look if you stored the toUpperCase() in a variable and return that variable?

File: code-challenges.test.js [hunk 3]
```diff
@@ -12,40 +12,124 @@
...
+
+//TODO --------------------2) Create a function that takes in a mixed data array and returns an array of only the REMAINDERS of the numbers when divided by 3.
+
+//TODO a) Create a test with an expect statement using the variables provided.
+
+describe('remainder3', () => {
```
Comment (SunkissedQueen) [inline code-challenges.test.js:54]: good

File: code-challenges.test.js [hunk 4]
```diff
@@ -12,40 +12,124 @@
...
+//  return array.filter(value => typeof value === 'number').map(value => value % 3)
+// }
+//* PASS  ./code-challenges.test.js
+//* aboutChar
+//*   ✓ takes in an array of objects and returns an array with a sentence about each person with their name capitalized. (2 ms)
+//* remainder3
+//*   ✓ takes in a mixed data array and returns an array of only the REMAINDERS of the numbers when divided by 3
+
+//TODO Refactor
+//? We can get rid of some returns and parentheses and curly braces.
```
Comment (SunkissedQueen) [inline code-challenges.test.js:81]: Yes!

File: code-challenges.test.js [hunk 5]
```diff
@@ -12,40 +12,124 @@
...
+const remainder3 = array => array.filter(value => typeof value === 'number').map(value => value % 3)
+
+//? I'm not sure why the aboutChar function ran slower this time around
```
Comment (SunkissedQueen) [inline code-challenges.test.js:84]: Not sure about that either...maybe the chain of the same hof. Just an hypothesis

File: code-challenges.test.js [hunk 6]
```diff
@@ -12,40 +12,124 @@
...
+//* aboutChar
+//*   ✓ takes in an array of objects and returns an array with a sentence about each person with their name capitalized. (2 ms)
+//* remainder3
+//*   ✓ takes in a mixed data array and returns an array of only the REMAINDERS of the numbers when divided by 3 (1 ms)
+//* cubeSum
+//*   ✓ takes in an array of numbers and returns the sum of all the numbers cubed
+
+//TODO Refactor
+//? Just tidying up a bit
+const cubeSum = array => array.map(value => value ** 3).reduce((sum, current)=> sum += current, 0)
```
Comment (SunkissedQueen) [inline code-challenges.test.js:127]: Looks great

File: interview-questions.md [hunk 7]
```diff
@@ -8,43 +8,85 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
   Your answer:
+  > We could generate a migration that will create a file that we can add a few additional commands. The name of the foreign key would be cohort_id. The foreign key would belong in the Student model, because Cohort has many students and the foreign key is there to identify which cohort the various students are a part of. Once the migration file was generated, we can add the command add_column followed by which table and what the name of the column would be and what data type it would be. Once we have specified, we could do a db:migrate to update our table.
 
   Researched answer:
-
+  > Another way we can accomplish this without having to manually type in the migration file would be to run the command, rails g migration <nameOfWhatWeWantToAdd> column_name:column_type. After we can simply type rails db:migrate. There are plenty of other commands we have access to besides add_column in the migration file.
```
Comment (SunkissedQueen) [inline interview-questions.md:14]: Wonderful

File: interview-questions.md [hunk 8]
```diff
@@ -8,43 +8,85 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
 
 2. Which RESTful routes must always be passed params? Why?
 
   Your answer:
+  > The show, edit, update, and destroy need them because they need to know  specific entry they need to GET, PATCH, or DELETE. The show requires the primary key to GET a specific entry. The edit method will require a parameter to offer a form specific for the entry you wish to change. The update and destroy need params as well to know which entry needs to be changed.
 
   Researched answer:
-
+  > Since RESTful routes are simply the patterns for how we map HTTP verbs to CRUD actions, there can be a variety of different routes that arise.  To the best of my understanding, create and new do not require params because they are not referencing a specific entry. There is no need because once a new entry is created a unique key will be given and the already existing entries are not altered. However, as we continue to create more routes as we pass data between routes, there will be a need for params to specify what we wish to request. For the wildlife api challenge, we had to add a route to specify the range of dates we wanted to see sightings for.
```
Comment (SunkissedQueen) [inline interview-questions.md:34]: beautiful

File: interview-questions.md [hunk 9]
```diff
@@ -8,43 +8,85 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
 
 3. Name three rails generator commands. What is created by each?
 
   Your answer:
-
+  > We have rails generate migration, rails generate Model, rails generate resources. The migration is for making changes to the table. We have the option to add or remove columns, change data types. The Model is for the initial creation of the table. The resources will create the table as well as the routes required for the controller.
   Researched answer:
+  > Each generate command has a variety of other options that can give further customization in the generation step. For example, we can add more options like what columns we want to add and what data type they should be in the model straight from the terminal. There is another generate command that we did not cover and that is the scaffold. Generators in and of itself do not create anything, rather they invoke other generators to do the work. The scaffold command invokes erb, test_unit, and helper generators to create an entire rails app for us to flesh out.
```
Comment (SunkissedQueen) [inline interview-questions.md:42]: fabulous

File: interview-questions.md [hunk 10]
```diff
@@ -8,43 +8,85 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
 
 
 4. Consider the Rails routes below. What is the name of the controller method that would be called by each route? What action would each of the controller methods perform?
```
Comment (SunkissedQueen) [inline interview-questions.md:47]: You got this concept

File: interview-questions.md [hunk 11]
```diff
@@ -8,43 +8,85 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+> This maps to the update method, because the parameter of primary key 2 let's us know which entry we are updating.  
 
-action: "DELETE" location: /students/2      
+action: "DELETE" location: /students/2 
+> This maps to the destroy method, because of the action delete as well as the parameter of the primary key 2 letting us know which entry to delete.     
 
 
 
 5. As a developer, you are making an application to manage your to do list. Create 10 user stories that will help you get your application started. Read more about [user stories](https://www.atlassian.com/agile/project-management/user-stories).
+
```
Comment (SunkissedQueen) [inline interview-questions.md:72]: Great stories

File: rails-commenting.rb [hunk 12]
```diff
@@ -7,24 +7,29 @@
 # FILE: app/controller/blog_posts_controller.rb
```
Comment (SunkissedQueen) [inline rails-commenting.rb:8]: fantastic way to share your knowledge

**Review by SunkissedQueen — APPROVED** [review]
Reviewing your assessments are a treat, Don. You always go beyond the surface of the concepts. I thoroughly enjoy seeing your refactors and research. Keep allowing yourself these opportunities to learn and grow. Your showing up for yourself and having such quality work helps you and those around you. Thank you also for not only having a grasp on the concept but also having the willingness to share it with others. You are the MVP!

**Comment by don-yoon** [comment]
Thank you for the great feedback. I've grown a lot from all the help.

**Pull request merged**